### PR TITLE
chore: Adds `resource_policy` model conversions

### PR DESCRIPTION
**Pull request by EspenAlbert**
## Description

- `resource_policy` model conversion

Link to any related issue(s): CLOUDP-272606

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR. A migration guide must be created or updated if the new feature will go in a major version.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR. A migration guide must be created or updated.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [ ] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have read the [contributing guides](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/README.md)
- [ ] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [ ] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [ ] I have added any necessary documentation (if appropriate)
- [ ] I have run make fmt and formatted my code
- [ ] If changes include deprecations or removals I have added appropriate changelog entries.
- [ ] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

### CODE REVIEW

File: internal/service/resourcepolicy/model.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	}
+	listType, diags := types.ListValueFrom(ctx, PolicyObjectType, tfModels)
+	if diags.HasError() {
+		return nil, diags
+	}
+	return &listType, nil
+}
+
+func NewTFPoliciesModelToSDK(ctx context.Context, input types.List) (*[]admin.ApiAtlasPolicyCreate, diag.Diagnostics) {
+	var tfPolicies []TFPolicyModel
```
Comment (EspenAlbert) [inline internal/service/resourcepolicy/model.go]: Both CREATE/UPDATE models only need to convert the policies:

```golang
type ApiAtlasResourcePolicyCreate struct {
	// Human-readable label that describes the atlas resource policy.
	Name *string `json:"name,omitempty"`
	// List of policies that make up the atlas resource policy.
	Policies *[]ApiAtlasPolicyCreate `json:"policies,omitempty"`
}
type ApiAtlasResourcePolicyEdit struct {
	// Human-readable label that describes the atlas resource policy.
	Name *string `json:"name,omitempty"`
	// List of policies that make up the atlas resource policy.
	Policies *[]ApiAtlasPolicyCreate `json:"policies,omitempty"`
}
```

File: internal/service/resourcepolicy/model.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+		LastUpdatedDate:   types.StringPointerValue(conversion.TimePtrToStringPtr(input.LastUpdatedDate)),
+		Name:              types.StringPointerValue(input.Name),
+		OrgID:             types.StringPointerValue(input.OrgId),
+		Policies:          *policies,
+		Version:           types.StringPointerValue(input.Version),
+	}, nil
+}
+
+func NewUserMetadataObjectType(ctx context.Context, input *admin.ApiAtlasUserMetadata) (*types.Object, diag.Diagnostics) {
+	if input == nil {
```
Comment (lantoli) [inline internal/service/resourcepolicy/model.go]: is it ok to call NewUserMetadataObjectType with nil input or it should be considered a logical error? same for NewPolicyObjectType

File: internal/service/resourcepolicy/model.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+		LastUpdatedByUser: *lastUpdatedByUser,
+		LastUpdatedDate:   types.StringPointerValue(conversion.TimePtrToStringPtr(input.LastUpdatedDate)),
+		Name:              types.StringPointerValue(input.Name),
+		OrgID:             types.StringPointerValue(input.OrgId),
+		Policies:          *policies,
+		Version:           types.StringPointerValue(input.Version),
+	}, nil
+}
+
+func NewUserMetadataObjectType(ctx context.Context, input *admin.ApiAtlasUserMetadata) (*types.Object, diag.Diagnostics) {
```
Comment (lantoli) [inline internal/service/resourcepolicy/model.go]: an alternative would be to have *diag.Diagnostics as param and add errors there

File: internal/service/resourcepolicy/model_test.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+	}
+
+	for testName, tc := range testCases {
+		t.Run(testName, func(t *testing.T) {
+			SDKModel := parseSDKModel(t, tc.SDKRespJSON)
+			ctx := context.Background()
+			resultModel, diags := resourcepolicy.NewTFResourcePolicyModel(ctx, &SDKModel)
+			assertDiagsOK(t, diags)
+
+			asserter := assert.New(t)
```
Comment (oarbusi) [inline internal/service/resourcepolicy/model_test.go]: why is this needed? wouldn't `assert.WhateverMethod()` work?

**Review by lantoli — APPROVED** [review]

File: internal/service/resourcepolicy/testdata/policy_clusterForbidCloudProvider.json
Comment (oarbusi) [inline internal/service/resourcepolicy/testdata/policy_clusterForbidCloudProvider.json:1]: this is a nice idea!

**Review by oarbusi — APPROVED** [review]
LGTM overall

File: internal/service/resourcepolicy/model.go (hunk 3 above)
Comment (lantoli) [reply internal/service/resourcepolicy/model.go]: curious to see how this function will be used

File: internal/service/resourcepolicy/testdata/policy_clusterForbidCloudProvider.json
Comment (lantoli) [reply internal/service/resourcepolicy/testdata/policy_clusterForbidCloudProvider.json:1]: I like it too. One small con is that data is not in the test so you need to look at both places to check what is going on in the test

Comment (oarbusi) [reply internal/service/resourcepolicy/testdata/policy_clusterForbidCloudProvider.json:1]: Agreed, probably this approach is more useful in cases where the json object is quite big and has a lot of attributes. For smaller ones, creating directly the SDK object might be better.

File: internal/service/resourcepolicy/model_test.go (hunk 4 above)
Comment (EspenAlbert) [reply internal/service/resourcepolicy/model_test.go]: We use this in other places too [ex1](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/CLOUDP-272606_rp_model_conversions/internal/testutil/acc/attribute_checks_test.go#L62), [ex2](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/CLOUDP-272606_rp_model_conversions/internal/testutil/mig/test_case_test.go#L28).
Just makes each of the assert calls a bit simpler

File: internal/service/resourcepolicy/model.go (hunk 3 above)
Comment (EspenAlbert) [reply internal/service/resourcepolicy/model.go]: It is used on line 13+17.
Why is it public you may ask?
1. Makes it easier to unit test
2. When generating the code, we don't know what will be usages. 
3. 🤔 Don't see why not?

WDYT?

File: internal/service/resourcepolicy/model.go (hunk 2 above)
Comment (EspenAlbert) [reply internal/service/resourcepolicy/model.go]: It is ok. Added a test case in ac5d66c7237c050dedbc0c5cd4cb613ee791a19a

File: internal/service/resourcepolicy/testdata/policy_clusterForbidCloudProvider.json
Comment (EspenAlbert) [reply internal/service/resourcepolicy/testdata/policy_clusterForbidCloudProvider.json:1]: IMO: I prefer this over typing it all out in the test. Especially, if you want to add more test cases.
Also, I think it is easier for auto-generation tools to work with an actual response, and then generate all the constants in the test.
Still, I'm struggling to find a nice way of initializing the TF models. As soon as there is some nesting, readability is terrible IMO

File: internal/service/resourcepolicy/model.go [hunk 5]
```diff
@@ -52,7 +53,8 @@ func NewUserMetadataObjectType(ctx context.Context, input *admin.ApiAtlasUserMet
 }
 
 func NewPolicyObjectType(ctx context.Context, input *[]admin.ApiAtlasPolicy) (*types.List, diag.Diagnostics) {
-	if input == nil {
+	var nilPointer *[]admin.ApiAtlasPolicy
```
Comment (EspenAlbert) [inline internal/service/resourcepolicy/model.go:48]: Thank you for spotting this @lantoli

File: internal/service/resourcepolicy/model.go (hunk 3 above)
Comment (EspenAlbert) [reply internal/service/resourcepolicy/model.go]: bfa7fa96fbac898a06cfc596d083d7cc75cc8ff1 @lantoli 
It looks cleaner now. Thank you for the suggestion ❤️
*Update: 013da09d2dae4d5e515d63beed1c1461f5d0ac3c even cleaner!
*Update2: d22eaa6ea9fbce9f69fc2855377875f798bc5918 last one?

File: internal/service/resourcepolicy/model_test.go [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+			userIDUpdate:    userIDUpdate,
+			userNameCreate:  userNameCreate,
+			userNameUpdate:  userNameUpdate,
+			createdDate:     createdDate,
+			lastUpdatedDate: lastUpdatedDate,
+			orgID:           orgID,
+			policyID:        policyID,
+			policyBody:      policyBody,
+			policies0ID:     policies0ID,
+			version:         version,
```
Comment (AgustinBettati) [inline internal/service/resourcepolicy/model_test.go]: q: is there a reason for not having a single expected TF model result here that can be asserted as we do in other unit tests? Currently each property is in the test case and then requires a particular assertion.

Comment (EspenAlbert) [reply internal/service/resourcepolicy/model_test.go]: Good callout.
I started that way to create the test "iteratively" to avoid having to develop the full TFModel.
However, I agree it is cleaner to do a single equal operator.
I refactored in 574799e672009523245228daa67a73537b8609f9 and during the refactoring, I felt the need for helpers, so I created 0313baed184cf1887884d834130bd51285b987c3
WDYT?

Comment (AgustinBettati) [reply internal/service/resourcepolicy/model_test.go]: Nice, thank you. 
I do still feel that `createTFModel` is doing very similar logic to the function we are actual testing `NewTFResourcePolicyModel`, which is why I like the idea of making the creation of the expected TF model as trivial as possible (ideally inlined in the test case directly). There might be some complications to what I am proposing so let me know your thoughts.

File: internal/service/resourcepolicy/resource_schema.go [hunk 7]
```diff
@@ -107,22 +107,29 @@ func ResourceSchema(ctx context.Context) schema.Schema {
...
-	Version           types.String `tfsdk:"version"`
+	CreatedByUser     types.Object    `tfsdk:"created_by_user"`
+	CreatedDate       types.String    `tfsdk:"created_date"`
+	ID                types.String    `tfsdk:"id"`
+	LastUpdatedByUser types.Object    `tfsdk:"last_updated_by_user"`
+	LastUpdatedDate   types.String    `tfsdk:"last_updated_date"`
+	Name              types.String    `tfsdk:"name"`
+	OrgID             types.String    `tfsdk:"org_id"`
+	Version           types.String    `tfsdk:"version"`
+	Policies          []TFPolicyModel `tfsdk:"policies"`
```
Comment (EspenAlbert) [inline internal/service/resourcepolicy/resource_schema.go:118]: @AgustinBettati thank you for showing how this is possible!

File: internal/service/resourcepolicy/model_test.go (hunk 6 above)
Comment (EspenAlbert) [reply internal/service/resourcepolicy/model_test.go]: Thanks for the conversation. Led to improvements.
I choose to keep the `createTFModel` as it simplifies distinguishing the test cases.

File: internal/service/resourcepolicy/model.go (hunk 5 above)
Comment (lantoli) [reply internal/service/resourcepolicy/model.go:48]: @EspenAlbert i was thinking more about this and I don't think this applies here, as the problem is only when using interfaces, and that's not the case here.  
I think nilPointer and nil will be the same in this case

File: internal/testutil/unit/fw_model.go [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+	ctx := context.Background()
+	list, diags := types.ListValueFrom(ctx, elementType, tfList)
+	AssertDiagsOK(t, diags)
+	return list
+}
+
+func AssertDiagsOK(t *testing.T, diags diag.Diagnostics) {
+	t.Helper()
+	if diags.HasError() {
+		t.Errorf("unexpected errors found: %s", diags.Errors()[0].Summary())
```
Comment (lantoli) [inline internal/testutil/unit/fw_model.go:30]: @EspenAlbert what about showing all errors with +v?  e.g.: https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/internal/provider/provider_test.go#L67
